### ISSUE TITLE
application not able to delete from argocd UI .It got stuck

### ISSUE THREAD

**Issue by Srivalle** [argoproj/argo-cd#12493]
### Discussed in https://github.com/argoproj/argo-cd/discussions/12492

<div type='discussions-op-text'>

<sup>Originally posted by **Srivalle** February 16, 2023</sup>
I am not able to delete the application from argocd ui and also from cli. 
I am getting the following error 
FATA[0010] rpc error: code = InvalidArgument desc = error patching application with finalizers: Application.argoproj.io "XXXX" is invalid: metadata.finalizers: Forbidden: no new finalizers can be added if the object is being deleted, found new finalizers []string{"resources-finalizer.argocd.argoproj.io"} . 
I am using 2.5.7 version of arogcd . Anyone could you please help here . </div>

**Comment by todaywasawesome** [comment]
@Srivalle I haven't seen this error when deleting an application however I have had the finalizer get stuck before. You can use this to force remove the finalizer
```kubectl patch some-resource/some-name \
    --type json \
    --patch='[ { "op": "remove", "path": "/metadata/finalizers" } ]'
```

Just change the resource to your application, however, you should make sure everything has been deleted. For example, are all the resources this application manages already deleted? I suspect it may be stuck there on some other finalizer.

**Comment by Srivalle** [comment]
In Argocd i have deleted the all the apps but still i am seeing the application is running in eks cluster. Even i deleted the namespace but still when i do kubectl get pods -n namespace i am seeing the pod and services  . What is the issue here . why it is reflecting in eks when we delete from UI

**Comment by Srivalle** [comment]
kubectl patch some-resource/some-name \
    --type json \
    --patch='[ { "op": "remove", "path": "/metadata/finalizers" } ]'
with this command can we delete the deployment, serivice as well ?

**Comment by todaywasawesome** [comment]
@Srivalle yes, but you should be somewhat careful with it. I suggest reading this resource https://martinheinz.dev/blog/74

**Comment by aikijitsu** [comment]
delete the pod

**Comment by ebuildy** [comment]
This is a very good question, in my opinion, 

We should add an option ""argocd app delete --force toto"" to the argocd-cli to remove the finalizers.

Or the finalizers is buggy ?

**Comment by ebuildy** [comment]
And a small tip, to remove finalizer of multiple argocd applications:

```
kubectl get applications  -o=jsonpath='{range .items[?(@.status.health.status=="Unknown")]}{.metadata.name}{"\n"}' | xargs -I {} kubectl patch application {}  --type=json -p='[{"op": "remove", "path": "/metadata/finalizers"}]'
```

**Comment by lyhuuloi** [comment]
Basically, if an Application or an ApplicationSet is stuck while deleting. It means it needs to wait for a response from "finalizers". So, the solution is to remove the "finalizers" from JSON.

![image](https://github.com/argoproj/argo-cd/assets/6936722/c2bca05c-a527-4068-9975-f44ff1340717)

**Comment by ebuildy** [comment]
This exactly what we are talking about ^^

BUT this is a bug from the argocd finalizers, it should not block if the app is on error.

**Comment by mario-e** [comment]
Adding my app manifest to the repo target referenced by the respective app results in stuck deletion for me as well.

AFAIK the idea of finalizers is to keep app around until all sub-components have been deleted, after which the controller should remove the finalizer for final clean-up. However in my case one of the sub-components is the app itself, which is kept around by the finalizer it seems - catch 22.

Not sure how to fix this or if there's something wrong w/ my architecture.

**Comment by mario-e** [comment]
manually removing label `argocd.argoproj.io/instance`/`app.kubernetes.io/instance` from app added by argocd unblocks the deletion

**Comment by ryanisnan** [comment]
I don't understand why this ticket was closed. We too have also ran into this issue, and the same fix works to address the immediate problem. This still leaves an open question though, why is this happening? Why can't ArgoCD delete resources it manages effectively?

**Comment by iamraj007** [comment]
Do we have a fix here or way to force delete that someone got it working?

**Comment by mrclrchtr** [comment]
I also think that should be reopened... It is not intuitive to have to do something like the workaround. It took me hours to figure that out.

**Comment by todaywasawesome** [comment]
@ryanisnan @iamraj007 @mrclrchtr The most common cause for this is from Kubernetes itself, not from Argo CD. Finalizers getting stuck in Kubernetes is a common issue. In the case of Argo, when we delete an application it needs to delete all managed resources and sometimes Kubernetes doesn't report the status of these items so the finalizer gets stuck. 

I most commonly see this because a pod's deleted status is missing. This is what happens, we tell the resource to delete and for whatever reason kubelet loses track of the running containers on the node and can't verify their deletion. If you did a `kubectl delete pod [podname] --force`, it's possible the container will actually continue running and taking up resources on the node. Generally, Argo CD, is just waiting for all components to be confirmed deleted before releasing the finalizer. 

If you are sure all resources have been deleted, and that nothing is waiting, then please share logs to show the application is stuck, as that would be a Argo CD issue. 

Otherwise, the fastest way to resolve this is to just remove the finalizer and let the resource delete with a command like this:

```
kubectl patch some-resource/some-name \
    --type json \
    --patch='[ { "op": "remove", "path": "/metadata/finalizers" } ]'
```

[Also please read the proper warnings on removing finalizers](https://martinheinz.dev/blog/74).

**Comment by cacaocal** [comment]
[root@YDY-CAN01-A25 crd]# argocd app list
NAME          CLUSTER                      NAMESPACE  PROJECT  STATUS   HEALTH   SYNCPOLICY  CONDITIONS                                         REPO                                          PATH                 TARGET
argocd/test2    mbb        wewv     Unknown  Unknown  <none>      InvalidSpecError,ComparisonError(2),DeletionError  https://gitee.com/quyi223/argocd-example.git  guestbook            master
argocd/test3   mbn        wewv     Unknown  Unknown  <none>      ComparisonError(2),DeletionError,InvalidSpecError  https://gitee.com/quyi223/argocd-example.git  kustomize-guestbook  master
[root@YDY-CAN01-A25 crd]# 
[root@YDY-CAN01-A25 crd]# 
[root@YDY-CAN01-A25 crd]# argocd app delete test2 -y 
FATA[0000] rpc error: code = InvalidArgument desc = error patching application with finalizers: Application.argoproj.io "test2" is invalid: metadata.finalizers: Forbidden: no new finalizers can be added if the object is being deleted, found new finalizers []string{"resources-finalizer.argocd.argoproj.io"} 


hi argcd ui and cli all not force delete, how to delete

**Comment by barrykaplan** [comment]
what exactly does `some-resource/some-name ` refer to? 

I have deleted all k8s resources.
However `argocd app list` still shows the zombie app as STATUS=OutOfSync and HEALTH=Missing

`argocd app delete <name>` says its deleted but still there

`argocd app terminate-op <name>` did finally get rid the app

**Comment by corneliusroemer** [comment]
It's frustrating, I also hit this issue and it required CLI intervention - wasn't doable from admin panel which is annoying.

```
kubectl patch app/APPNAME \
    --type json \
    --patch='[ { "op": "remove", "path": "/metadata/finalizers" } ]'
```

Replace APPNAME with the app of your app.

**Comment by qs-wang** [comment]
> what exactly does `some-resource/some-name ` refer to?
> 
> I have deleted all k8s resources. However `argocd app list` still shows the zombie app as STATUS=OutOfSync and HEALTH=Missing
> 
> `argocd app delete <name>` says its deleted but still there
> 
> `argocd app terminate-op <name>` did finally get rid the app

Refer to the application in this case
```
kubectl patch application yourapp  \
  --type=json -p='[{"op": "remove", "path": "/metadata/finalizers"}]' -n argocd
```

**Comment by aiell0** [comment]
There definitely needs to be a `force delete` option that just runs this patch command against the finalizers. Just ran into this issue today.

**Comment by alen-z** [comment]
Same issue. Deployments from PR generator deploy Application resource that, if stuck in `OutOfSync` state fro whatever reason and PR branch is deleted, combination of trying to sync from non-existing branch and triggered deletion process does not play well together. Argo is not deleting all deployed resources by Application, Application is not removed because it has finalizers, ApplicationSet not removing Application since it has finalizers.

We need some force cleanup method if generated Application and/or deployed resources are stuck and source is not available anymore.

- Reopen.

**Comment by pydevops** [comment]
The following is a simple shell script I use for removing the argocd based on the previous post. 
```
#!/usr/bin/env bash
APPS=$(kubectl -n argocd get app -o jsonpath='{range .items[*]}{.metadata.name}{"\n"}{end}')
for app in $APPS
do
 echo "patch $app 's finalizer"
 kubectl patch app/$app \
    --type json \
    --patch='[ { "op": "remove", "path": "/metadata/finalizers" } ]'
done
```

**Comment by sfratini** [comment]
Just came to mention another item. I had an "app of apps" and the children apps were stuck. I was not able to delete the children apps but I was able to delete the main one. When I recreated it, the old apps were still showing again as pending deletion. I saw the manifest had the finalizer annotation so I thought I could try to manually change it. When I hit save, it automatically said "Resource not found" and refreshed the state of the app, which is not pending so I can continue the deployment. This could just be solved with a way for us to refresh that state from the UI.

**Comment by Malachyiii** [comment]
This keeps getting closed, but it has not been solved. 

I just want to be explicitly clear: Is the opinion of the ArgoCD maintainers that this is **unsolvable** and the **only solution** is command line work arounds? There's no possibility AT ALL of a force delete option for the UI?

That is the only reason I can see for why it keeps getting closed without an actual fix. I'm actively dealing with a zombie app right now that has the same issue.

**Comment by cyril-accenta** [comment]
I'm still experiencing this bug. Deleting an app results in a stuck finalizer. Removing the finalizer immediately resolves the issue, but I agree that this shouldn't be necessary

**Comment by corneliusroemer** [comment]
> This keeps getting closed, but it has not been solved.

I'm not seeing repeated closures? It was closed 2y ago and never reopened. Maybe someone simply needs to make a new issue
